### DISCUSSION TITLE
Using redirect / notFound in getServerSideProps throws  "Did you forget to add a `return`?"

### DISCUSSION THREAD

**Discussion by ramiant** [vercel/next.js#51352] [Help] [answered]
### Summary

I'm trying to redirect directly from server instead of using client router.
Using the api described here: [https://nextjs.org/docs/pages/api-reference/functions/get-server-side-props#redirect](https://nextjs.org/docs/pages/api-reference/functions/get-server-side-props#redirect)

```
export const getServerSideProps = gsspWithNonce(async ({ query, locale, resolvedUrl }) => {
  if (!query.platform) {
    return {
      redirect: {
        permanent: false,
        destination: resolvedUrl + `&platform=web`,
      },
    };
  }

  return {
      props: {}
  };
});
```

**Expected Behavior**
Page is redirected to new url with concated query param.

**Actual Behavior**
```error - Error: Your `getServerSideProps` function did not return an object. Did you forget to add a `return`?```

Is the behavior implicitly implemented in nextjs or do I have to do something extra for it to work?

**Note**
This happens also if I try returning the `notFound: boolean` .

### Additional information

```js
Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.4.0: Mon Mar  6 20:59:28 PST 2023; root:xnu-8796.101.5~3/RELEASE_ARM64_T6000
    Binaries:
      Node: 16.17.0
      npm: 9.4.0
      Yarn: 1.22.19
      pnpm: N/A
    Relevant packages:
      next: 13.2.4
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
```

**Comment by icyJoseph** [comment]
I'd guess the problem is `gsspWithNonce`, what happens if you remove it? or could you share its implementation?

**Reply by ramiant** [reply]
Implementation of `gsspWithNonce` is described here [https://www.npmjs.com/package/@next-safe/middleware](https://www.npmjs.com/package/@next-safe/middleware)

**Reply by icyJoseph** [reply, marked as answer]
Japp, https://github.com/nibtime/next-safe-middleware/blob/c4075709fd8598ff3af27bafd25e0f298509dac5/packages/next-safe-middleware/src/document/NextPageContext/index.ts#L5

```ts
export function gsspWithNonce<
  P extends { [key: string]: any } = { [key: string]: any },
  Q extends ParsedUrlQuery = ParsedUrlQuery,
  D extends PreviewData = PreviewData
>(
  getServerSideProps: GetServerSideProps<P, Q, D>
): GetServerSideProps<P & { nonce: string }, Q, D> {
  return async (ctx) => {
    const gsspResult = await getServerSideProps(ctx);
    if ("props" in gsspResult) {
      const nonce = getCreateCtxNonceIdempotent(ctx);
      const props = await gsspResult.props;
      return { props: { ...props, nonce } };
    }
  };
}
```

It doesn't support returning anything that doesn't have props, otherwise it returns `undefined`. 

Even if you include empty props in your redirect branch, it'll force return props, without redirect.

I guess you could refactor:


```tsx
export const getServerSideProps = async (context) => {
  const { query, locale, resolvedUrl } = context;
  if (!query.platform) {
    return {
      redirect: {
        permanent: false,
        destination: resolvedUrl + `&platform=web`,
      },
    };
  }

  return gsspWithNonce(ctx => {
      return {props: {}}
   })(context);
}
```

Or something like that, or just use your own nonce, or ask them to fix their function. It should be:

```tsx
export function gsspWithNonce<
  P extends { [key: string]: any } = { [key: string]: any },
  Q extends ParsedUrlQuery = ParsedUrlQuery,
  D extends PreviewData = PreviewData
>(
  getServerSideProps: GetServerSideProps<P, Q, D>
): GetServerSideProps<P & { nonce: string }, Q, D> {
  return async (ctx) => {
    const gsspResult = await getServerSideProps(ctx);
    if ("props" in gsspResult) {
      const nonce = getCreateCtxNonceIdempotent(ctx);
      const props = await gsspResult.props;
      return { props: { ...props, nonce } };
    }

    return gsspResult; // this is missing
  };
}
```

**Reply by icyJoseph** [reply]
There's a PR from a couple of weeks back, https://github.com/nibtime/next-safe-middleware/pull/95, but nobody seems to give it attention.

**Reply by ramiant** [reply]
Yes it seems the issue comes from the wrapper function source code. Hopefully someone will take the PR at one time. The refactoring suggestion works though so thank you, I appreciate your insight.
